### ISSUE TITLE
Submodule params are not set properly

### ISSUE THREAD

**Issue by MichaelKonobeev** [facebookresearch/higher#19]
The following example leads to `TypeError: 'NoneType' object is not subscriptable` raised [here](https://github.com/facebookresearch/higher/blob/master/higher/patch.py#L287):
```
class Module(nn.Module):
    def __init__(self):
        super().__init__()
        self.submodule = nn.Linear(3, 4)

    def forward(self, inputs):
        return self.submodule(inputs)

module = Module()
fmodule = monkeypatch(module)

print(fmodule.submodule(torch.randn(2, 3)).shape)
```

It seems like this is happening because `params_box` has single value of `None` in it as set [here](https://github.com/facebookresearch/higher/blob/master/higher/patch.py#L328) unless the wrapped module's forward function is called which causes it to be changed [here](https://github.com/facebookresearch/higher/blob/master/higher/patch.py#L348). Calling the module on an input and then the submodule does not lead to this issue, that is the following works properly:
```
inputs = torch.randn(2, 3)
fmodule(inputs)
fmodule.submodule(inputs) # OK
```

Modifying `params_box` in the `_update_params` as in [here](https://github.com/MichaelKonobeev/higher/commit/6966f9ec1151f53f2dd2696e59cbcd21d0580046) fixes the issue but is this a proper fix? I am not sure what [`_update_patched_params`](https://github.com/MichaelKonobeev/higher/blob/submodule-params-bug-fix/higher/patch.py#L297) is doing because the parameters are also set in [`patched_forward`](https://github.com/MichaelKonobeev/higher/blob/submodule-params-bug-fix/higher/patch.py#L284) each time it's called.

**Comment by egrefen** [comment]
Hello. I've returned from leave and allocated some time to look into this issue over the next two weeks. Hopefully we'll make some progress and report back, or come back to you with questions.

**Comment by egrefen** [comment]
This isn't a bug per se, but rather a limitation of how we implement the recursive monkey patching. The assumption was that people would not call submodules directly. I don't think this is a good assumption to make, in hindsight, and I have some ideas on how to fix things. 

I will try to plan out my solution and the required tests, and also study how to generalize your proposed fixes to ensure that the approach is robust to all use cases. 

To manage expectations, I will probably not have time to do this until next week, but I'm confident I  can resolve the issue by the end of that week (famous last words). Let me know if this is more time sensitive for your project(s) and I can help with a quick workaround.

**Comment by Franklin-Yao** [comment]
I also have this issue.

**Comment by egrefen** [comment]
@MichaelKonobeev @Franklin-Yao I'm sorry this issue has not been addressed. There's been a bit of delay on all fronts with covid + ICML submissions. I'm the only person working on this project so I'm afraid it's been a bit slow moving. I'll try to address this in the coming weeks.

**Comment by egrefen** [comment]
Fixed in #47. Let me know if you still have issues!
